appliances/Harbor/appliance.sh

```diff
@@ -0,0 +1,304 @@
+    {
+        if ($1 == "harbor_admin_password:") {
+            print "harbor_admin_password:", admin_pwd
+        } else if ($1 == "password:") {
+            print "  password:", db_pwd
+        } else {
+            print
+        }
+    }' /root/harbor/harbor.yml
+    msg info "Configuration phase finished"
+}
+
+service_bootstrap() { # Running Harbor installer => will generate and run docker-compose
+    msg info "Starting bootstrap..."
+    /root/harbor/install.sh
+    if [ $? -ne 0 ]; then
+	    msg error "Harbor installation script failed, aborting..."
+        exit 1
+    else
+        msg info "Harbor installation script finished successfully. Waiting 60s before final health checks..."
+    fi
+    sleep 60
+    msg info "Running final health checks..."
+    wait_for_docker_containers
+    cleanup_installation
+    msg info "Bootstrap phase finished"
+
+}
+
+### Function Definitions
+
+check_internet_access() {
+    # Ping Google's public DNS server
+    if ping -c 1 8.8.8.8 &> /dev/null; then
+        msg info "Internet access OK"
+        return 0
+    else
+        msg error "The VM does not have internet access. Aborting Harbor deployment..."
+        exit 1
+    fi
+}
+
+install_requirements(){
+    apt update && apt install openssl ca-certificates curl gnupg docker.io docker-compose -y
+    if [ $? -eq 0 ]; then
+	    msg info "Dependencies installation finished successfully. Running extra checks..."
+    else
+        msg error "Dependencies installation failed. Aborting..."
+    fi
+    check_docker
+    check_docker_compose
+}
+
+check_docker() {
+    msg info "Checking if Docker is installed and running..."
+    if ! command -v docker &> /dev/null; then
+        msg error "Docker could not be found, Docker installation failed..."
+    else
+        msg info "Docker is installed, ensuring it is running..."
+        if ! systemctl is-active --quiet docker; then
+            msg warning "Docker is not running, attempting to start Docker..."
+            systemctl start docker
+            systemctl enable docker
+            sleep 5
+        fi
+        if ! systemctl is-active --quiet docker; then
+            msg error "Failed to start Docker"
+            exit 1
+        else
+            msg info "Docker is running"
+            return 0
+        fi
+    fi
+}
+
+check_docker_compose() {
+    msg info "Checking if docker-compose is installed..."
+    if ! command -v docker-compose &> /dev/null; then
+        msg error "docker-compose could not be found, aborting..."
+        exit 1
+    else
+        msg info "docker-compose is installed"
+        return 0
+    fi
+}
+
+
+
+generate_ssl_certs() {
+    if [ -z "${HARBOR_HOSTNAME}" ]; then
+        msg info "HARBOR_HOSTNAME is not set or is empty. Generating certs with 'example.com'"
+        HARBOR_HOSTNAME=example.com
+    fi
+    openssl req -new -newkey rsa:2048 -sha256 -days 365 -nodes -x509 \
```

Comment:
So since we require Internet access why not use ACME certbot here?

Reply:
I thought about it when writing the code. 


The appliance will have outbound internet access, but will probably not be reachable from the public internet. Let's Encrypt CA needs to validate the provided domain in order to generate the cert. The three existing methods for validating the domain (HTTP challenge, TLS challenge, DNS challenge) do not work with blocked inbound traffic and/or private local DNS.